### PR TITLE
Convert BaseDom and its derived classes to use initializers

### PR DESCRIPTION
**Pull request by benharsh**
This PR converts BaseDom and all of its derived classes to use initializers instead of constructors. This is accomplished by either adding a "use default init" pragma or by converting an existing constructor to an initializer.

A small compiler change is made to fix the name of DefaultRectangularDom so that users just see "domain". This functionality existed before, but only for constructors.

Future work:
- Allow field default value to be call on the result of another call (DefaultAssociativeDom)
- Rework BaseDom* classes in ChapelDistribution to use explicit initializers, and stop using named args in derived class' super.init.

Testing:
- [x] full local + futures
- [x] full gasnet

### CODE REVIEW

**Comment by benharsh** [comment]
As of this PR, down to 222 failures for ``--force-initializers``. The improvements likely came from already-merged PRs developed in support of this PR.

File: compiler/resolution/resolveFunction.cpp [hunk 1]
```diff
@@ -853,7 +855,7 @@ static void fixTypeNames(AggregateType* at) {
 
     at->symbol->name = astr("[", domainType, "] ", eltType);
 
-  } else if (from != NULL && strcmp(from->symbol->name, domName) == 0) {
+  } else if (strncmp(at->symbol->name, domName, domNameLen) == 0) {
     at->symbol->name = astr("domain", at->symbol->name + strlen(domName));
```
[the remark this answers is not in the document]
Comment (benharsh) [reply compiler/resolution/resolveFunction.cpp]: Good catch! I'll make the change momentarily.

**Comment by Tshimanga** [comment]
@benharsh would this affect my issue here https://github.com/chapel-lang/chapel/issues/9167 ?

**Comment by benharsh** [comment]
I'm not sure, I haven't had the chance to think much about that issue (yet). I suspect the cause is more likely to be related to the issues Brad and Michael discussed, which this PR doesn't address.

**Pull request merged**